### PR TITLE
removing webpki::Error as return type

### PR DESCRIPTION
**Pull request by Herbert-Karl**
hi,
following I propose a solution for issue #839. As I am new to Rust and the codebase of rustls, I don't have much confidence, that the change doesn't break API or introduces side effects.
I have run the existing tests and no errors were reported. I haven't fuzzed the changed codebase or did any manual testing.

### CODE REVIEW

**Review by ctz — DISMISSED** [review]
Thanks! Going to hold this one for the next release, as it's a breaking API change.

File: rustls/src/anchors.rs [hunk 1]
```diff
@@ -82,8 +83,9 @@ impl RootCertStore {
     }
 
     /// Add a single DER-encoded certificate to the store.
-    pub fn add(&mut self, der: &key::Certificate) -> Result<(), webpki::Error> {
-        let ta = webpki::TrustAnchor::try_from_cert_der(&der.0)?;
+    pub fn add(&mut self, der: &key::Certificate) -> Result<(), Error> {
+        let ta = webpki::TrustAnchor::try_from_cert_der(&der.0)
+            .map_err(|e| Error::InvalidCertificateData(format!("{}", e)))?;
```
Comment (briansmith) [inline rustls/src/anchors.rs:88]: I suggest we use a different, new, error variant: `InvalidRootCertData`. `InvalidCertificateData` should only be returned during certificate validation.

**Review by briansmith — CHANGES_REQUESTED** [review]
Thanks for working on this. I just have one suggestion.

File: rustls/src/anchors.rs (hunk 1 above)
Comment (Herbert-Karl) [reply rustls/src/anchors.rs:88]: i can understand that though
based on the description and error cases in the "try_from_cert_der" function, an error variant called "FailtedToDecodeDER" seems more appropriate, as the function doesn't actually validate the certificate, only decodes it

Comment (briansmith) [reply rustls/src/anchors.rs:88]: > i can understand that though
> based on the description and error cases in the "try_from_cert_der" function, an error variant called "FailtedToDecodeDER" seems more appropriate, as the function doesn't actually validate the certificate, only decodes it

The problem I see with `FailedtoDecodeDER` is that it doesn't provide any context to the user about what was trying to be decoded. Also, I think `InvalidRootCertData` could be used in the future for any kind of problem with the root cert, not just problems with the ASN.1 encoding.

Comment (Herbert-Karl) [reply rustls/src/anchors.rs:88]: You are right, `FailedToDecodeDER` by name lacks context, in which the error occurred. I don't like that `InvalidRootCertData` suggests something (a validation of the root certificate) which doesn't actually happen.
What do you think about naming the error `FailedToDecodeRootCertData`?

**Review by briansmith — APPROVED** [review]

**Comment by ctz** [comment]
Thanks for the contribution here -- recently we've been trying to move away from errors that aren't machine readable because they squashed the cause into a string -- see eg. #1162. This has been fixed in the progress towards that -- see a777077.

**Pull request closed without merging**